### PR TITLE
kv/rocksdb: New (Soon to be) Binned LRU Cache for RocksDB

### PR DESCRIPTION
**Pull request by markhpc**
This PR implements a modified LRU block cache for RockSDB based on RocksDB's internal LRU block cache.  Currently the primary differences between this cache and RocksDB's is that we expose functionality not present in RocksDB's generic cache interface.  This includes the ability to Set the high priority pool ratio and get the current usage of the high priority pool.  We also explicitly disable the promotion of data blocks into the high priority pool, while indexes/filters are still allowed to reside there.  Versus the RocksDB's LRU cache, this will result in higher priority given to indexes and filters over data that has a read hit and also allow us to better account for how much of the cache is currently being dedicated to indexes and filters (though indexes/filters can still fall out of the high priority pool if they grow too quickly).

Priority binning is not yet implemented though this PR puts us on track for being able to implement this without requiring upstream rocksdb PRs (The rocksdb folks indicated that they would like to limit the number of modifications to the existing lru_cache).

~~Several RocksDB utility classes are used in this PR for mutex handling, autovector and hashing, though we may be able to migrate this code to use existing ceph classes for these things over time.~~

An autovector utility class is used in this PR to optimize small vector operations.

### CODE REVIEW

**Comment by markhpc** [comment]
I'm going to clean up some of the RocksDB utility classes to shrink this PR down some.

**Comment by markhpc** [comment]
Ok, this feels better now.

All rocksdb util dependencies have been removed except for autovector, which is an optimized vector that does stack based allocations for small item counts.  We can optionally remove this dependency by switching to std::vector directly, though I figured it may be useful enough to merit a spot in common.

Other things changed:

- RocksDB Mutex -> std::mutex
- RocksDB "murmur like" hash -> CEPH_STR_HASH_RJENKINS
- Use alignas directly
- Use posix_memalign unconditionally since we don't need other architectures and gcc<5.  At a future point this can be migrated to use the mempool framework.

**Comment by liewegas** [comment]
```
2018-07-09T14:51:19.497 INFO:teuthology.orchestra.run.smithi047.stdout:The following packages have unmet dependencies:
2018-07-09T14:51:19.529 INFO:teuthology.orchestra.run.smithi078.stdout:Building dependency tree...
2018-07-09T14:51:19.529 INFO:teuthology.orchestra.run.smithi078.stdout:Reading state information...
2018-07-09T14:51:19.533 INFO:teuthology.orchestra.run.smithi047.stdout: ceph : Depends: ceph-mon (= 14.0.0-1164-g444699a-1xenial) but it is not going to be installed
2018-07-09T14:51:19.533 INFO:teuthology.orchestra.run.smithi047.stdout:        Depends: ceph-osd (= 14.0.0-1164-g444699a-1xenial) but it is not going to be installed
2018-07-09T14:51:19.533 INFO:teuthology.orchestra.run.smithi047.stdout: ceph-mds : Depends: ceph-base (= 14.0.0-1164-g444699a-1xenial) but it is not going to be installed
2018-07-09T14:51:19.533 INFO:teuthology.orchestra.run.smithi047.stdout: ceph-mgr : Depends: ceph-base (= 14.0.0-1164-g444699a-1xenial) but it is not going to be installed
2018-07-09T14:51:19.533 INFO:teuthology.orchestra.run.smithi047.stdout: ceph-test : Depends: libstdc++6 (>= 7) but 5.4.0-6ubuntu1~16.04.9 is to be installed
2018-07-09T14:51:19.549 INFO:teuthology.orchestra.run.smithi047.stderr:E: Unable to correct problems, you have held broken packages.
```
that last bit about libstdc++6 seems suspicious to me?  @tchaikov any ideas?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-07-09_14:35:36-rados-wip-sage4-testing-2018-07-06-1705-distro-basic-smithi/

**Comment by markhpc** [comment]
That's interesting.  libstdc++ 5.4 is the latest version in Xenial afaik:

https://packages.ubuntu.com/xenial/libstdc++6

However install-deps.sh pulls gcc7 from ubuntu-toolchain-r:

http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu

And specifically that should pull in:

http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu/pool/main/g/gcc-7/libstdc++6_7-20170407-0ubuntu2_amd64.deb

File: src/common/autovector.h [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+//     need to copy the arrays other than just swapping the pointers. In this
+//     case we'll just explicitly forbid these operations since they may
+//     lead users to make false assumption by thinking they are inexpensive
+//     operations.
+//
+// Naming style of public methods almost follows that of the STL's.
+namespace ceph {
+
+template <class T, size_t kSize = 8>
+class autovector {
```
Comment (cbodley) [inline src/common/autovector.h]: would [boost::container::small_vector](https://www.boost.org/doc/libs/1_67_0/doc/html/boost/container/small_vector.html) work for this?

Comment (markhpc) [reply src/common/autovector.h]: @bodley oh, I forgot about boost:container::small_vector.  Yes, I imagine it would!

**Comment by markhpc** [comment]
retest this please

**Comment by markhpc** [comment]
retest this please

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-08-08_18:17:37-rados-wip-sage3-testing-2018-08-08-1057-distro-basic-smithi/ (looks good)

File: src/kv/RocksDBStore.cc [hunk 2]
```diff
@@ -1306,21 +1312,6 @@ int64_t RocksDBStore::commit_cache_size()
...
-  }
-  dout(10) << __func__ << " old ratio: " 
-          << bbt_opts.block_cache->GetHighPriPoolRatio() << " new ratio: "
-          << ratio << dendl;
-  bbt_opts.block_cache->SetHighPriPoolRatio(ratio);
-  return 0;
-}
-
-=======
->>>>>>> src/kv: Initial import of a custom RocksDB cache.
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: can you squash this part into the earlier commit that introduces it?

File: src/common/autovector.h (hunk 1 above)
Comment (tchaikov) [reply src/common/autovector.h]: @markhpc i also think `boost::container::small_vector` might be a better choice here. could you reconsider this proposal?

Comment (markhpc) [reply src/common/autovector.h]: @tchaikov @cbodley Absolutely!  I've been a little hesitant to change it as part of this PR though as I already did a fair amount of testing using their existing autovector code and I'd like to make the change to small_vector as a separate future PR.  Does that seem reasonable?

File: src/kv/RocksDBStore.cc (hunk 2 above)
Comment (markhpc) [reply src/kv/RocksDBStore.cc]: ugh, that's irritating.  Let me see what I can do.

**Pull request merged**